app/src/ui/multi-commit-operation/choose-branch/base-choose-branch-dialog.tsx

```diff
@@ -65,85 +89,32 @@ interface IBaseChooseBranchDialogProps {
   readonly onDismissed: () => void
 }
 
-export interface IBaseChooseBranchDialogState {
-  /** The currently selected branch. */
+export interface IChooseBranchDialogProps extends IBaseChooseBranchDialogProps {
   readonly selectedBranch: Branch | null
+  readonly dialogTitle: string | JSX.Element | undefined
+  readonly submitButtonTooltip?: string
+  readonly canStartOperation: boolean
+  readonly start: () => void
+  readonly onSelectionChanged: (selectedBranch: Branch | null) => void
+}
 
+export interface IChooseBranchDialogState {
   /** The filter text to use in the branch selector */
   readonly filterText: string
-
-  /**
-   * A preview of the operation using the selected base branch to test whether the
-   * current branch will be cleanly applied.
-   */
-  readonly statusPreview: JSX.Element | null
 }
 
-export abstract class BaseChooseBranchDialog extends React.Component<
-  IBaseChooseBranchDialogProps,
-  IBaseChooseBranchDialogState
+export class ChooseBranchDialog extends React.Component<
+  IChooseBranchDialogProps,
+  IChooseBranchDialogState
 > {
-  protected abstract start: () => void
-
-  protected abstract canStart: () => boolean
-
-  protected abstract updateStatus: (branch: Branch) => Promise<void>
-
-  protected abstract getDialogTitle: (
-    branchName: string
-  ) => string | JSX.Element | undefined
-
-  protected abstract renderActionStatusIcon: () => JSX.Element | null
-
-  public constructor(props: IBaseChooseBranchDialogProps) {
+  public constructor(props: IChooseBranchDialogProps) {
     super(props)
 
-    const selectedBranch = this.resolveSelectedBranch()
-
     this.state = {
-      selectedBranch,
       filterText: '',
-      statusPreview: null,
     }
-  }
 
-  public componentDidMount() {
-    const { selectedBranch } = this.state
-    if (selectedBranch !== null) {
-      this.updateStatus(selectedBranch)
-    }
-  }
-
-  protected getSubmitButtonToolTip = (): string | undefined => {
-    return undefined
-  }
-
-  private onFilterTextChanged = (filterText: string) => {
-    this.setState({ filterText })
-  }
-
-  private onItemClick = (branch: Branch, source: ClickSource) => {
-    if (source.kind !== 'keyboard' || source.event.key !== 'Enter') {
-      return
-    }
-
-    source.event.preventDefault()
-
-    const { selectedBranch } = this.state
-
-    if (selectedBranch !== null && selectedBranch.name === branch.name) {
-      this.start()
-    }
-  }
-
-  protected onSelectionChanged = async (selectedBranch: Branch | null) => {
-    if (selectedBranch != null) {
-      this.setState({ selectedBranch })
-      return this.updateStatus(selectedBranch)
-    }
-
-    // return to empty state
-    this.setState({ selectedBranch })
+    this.props.onSelectionChanged(this.resolveSelectedBranch())
```

Comment:
Using callback props before a component is mounted is dangerous. Consumers should be able to assume that the callbacks passed to a component is only used after that component has mounted.


We should move this to `componentDidMount` and we should only call `onSelectionChanged` if the selection has actually changed (i.e. if the return value of `resolveSelectedBranch` differs from the `selectedBranch` prop).